.changeset/shiny-swans-relax.md

```diff
@@ -0,0 +1,80 @@
+Due to limitations of the TypeScript language, there was not a practical way to prevent less descriptive types, like `TreeNodeSchema` or `ImplicitFieldSchema` from being used where the type of a specific schema was intended.
+Code which does this will encounter several issues with tree APIs, and this change fixes some of those issues.
+This change  mainly fixes that `NodeFromSchema<TreeNodeSchema>` used to return `unknown` and now returns `TreeNode | TreeLeafValue`.
+
+This change by itself seems mostly harmless, as it just improves the precision of the typing in this one edge case.
+Unfortunately there are other typing bugs which complicate the situation, resulting in APIs for inserting data into the tree to also behave poorly when given non-specific types like `TreeNodeSchema`.
+These APIs include cases like `TreeView.initialize`.
+
+This incorrectly allowed some usage like taking a type erased schema and initial tree pair, and creating a view of type `TreeView<ImplicitFieldSchema>` then initializing a it.
+With the typing being partly fixed, some unsafe inputs are still allowed when trying to initialize such a view, but some are now prevented.
+
+While this use-case of modifying in code not strongly typed by the exact schema, was not intended to be supported,
+it did mostly work, and has some real use-cases (like tests looping over test data pairs of schema and tree data).
+To help mitigate the impact of this change, some experimental `@alpha` APis have been introduced.
+
+Before this change:
+
+```typescript
+import { TinyliciousClient } from "@fluidframework/tinylicious-client";
+import {
+	SchemaFactory,
+	SharedTree,
+	TreeViewConfiguration,
+	type TreeNodeSchema,
+} from "fluid-framework";
+
+// Create a ITree instance
+const tinyliciousClient = new TinyliciousClient();
+const { container } = await tinyliciousClient.createContainer({ initialObjects: {} }, "2");
+const tree = await container.create(SharedTree);
+
+const schemaFactory = new SchemaFactory("demo");
+
+// Bad: This loses the schema aware type information. `: TreeNodeSchema` should be omitted to preserve strong typing.
+const schema: TreeNodeSchema = schemaFactory.array(schemaFactory.number);
+const config = new TreeViewConfiguration({ schema });
+
+// This view is types as `TreeView<TreeNodeSchema>`, which does not work well since its missing the actual schema type information.
+const view = tree.viewWith(config);
+// Root is typed as `unknown` allowing invalid assignment operations.
+view.root = "invalid";
+view.root = {};
+// Since all assignments are allowed, valid ones still work:
+view.root = [];
+```
+
+After this change:
+
+
+```typescript
+// Root is now typed as `TreeNode | TreeLeafValue`, still allowing some invalid assignment operations.
+// In the future this should be prevented as well, since the type of the setter in this case should ne `never`
```

Comment:
We can't generate different getter and setter types via mapped types though - so, will it even be possible to have the setter be `never` but the getter remain `TreeNode | TreeLeafValue`? Maybe it is possible in this one case because it is the base case, and we don't need to use a mapped type for the top level?

Reply:
We use mapped types for the schema aware API for object nodes: it someone for some reason decides to construct the type that is an object node with one of its field schema set to ImplicitFIeldSchema, or something like that, we could handle it someday by not generating a setter, which is mostly the same as a never setter. That is possible.

Additionally if we someday support UnsafeUnknownSchema being used in ObjectNode field schema, that actually works fine since that gives you both a get and set of `TreeNode | TreeLeaftValue` which is also expressible.

This does lead into an interesting API alternative, which as it only changes the alpha parts shouldn't block this:
We could make UnsafeUnknownSchema a branded or wrapped form of TreeNodeSchema, which would make it reasonable to use when defining schema as a runtime value. That could pair decently with such fields taking in branded input so you would have to do something like `MyNode.myunsafeField = intoUnsafe(foo)`.

Personally I think thats worse than just not using UnsafeUnknownSchema in node schema, and forcing use of something like Tree.trySetField(node, field, value), or Tree.tryEditArray(...) or Tree.tryDetachOptional(...) for non schema aware editing if we want to support it. Then we just make fields without known schema types readonly.

Reply:
Interesting ideas...